File: 5-Problem-A-Day/318. Maximum Product of Word Lengths.java
class Solution {
    public int maxProduct(String[] words) {
        int n = words.length;
        int[] masks = new int[n];
        int[] lens = new int[n];
        
        int bitmask = 0;
        for (int i = 0; i < n; i++) {
            bitmask = 0;
            for (char ch: words[i].toCharArray()) {
                bitmask |= 1 << (int)(ch - 'a');
            }
            masks[i] = bitmask;
            lens[i] = words[i].length();
        }
        
        int maxVal = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if ((masks[i] & masks[j]) == 0)
                    maxVal = Math.max(maxVal, lens[i] * lens[j]);
            }
        }
        
        return maxVal;
    }
}

File: 2021/318. Maximum Product of Word Lengths.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        vector<int> counts(words.size(), 0);
        for (int i = 0; i < words.size(); i++) {
            for (char c : words[i]) {
                counts[i] |= (1 << (c - 'a'));
            }
        }
        int res = 0;
        for (int i = 0; i < words.size(); i++) {
            for (int j = i; j < words.size(); j++) {                            
                if ((counts[i] & counts[j]) == 0) {
                    res = max(res, (int)(words[i].length() * words[j].length()));
                }
            }
        }
        return res;
    }
};